### ISSUE TITLE
A question about victorialogs generating a stream ID

### ISSUE THREAD

**Issue by acyouzi** [VictoriaMetrics/VictoriaMetrics#5532]
### Is your question request related to a specific component?

A question about victorialogs

### Describe the question in detail

        func (lr *LogRows) MustAdd(tenantID TenantID, timestamp int64, fields []Field) {
                // Compose StreamTags from fields according to lr.streamFields
                sfs := lr.streamFields
                st := GetStreamTags()
                for i := range fields {
                        f := &fields[i]
                        if _, ok := sfs[f.Name]; ok {
                                st.Add(f.Name, f.Value)
                        }
                }
                // Marshal StreamTags
                bb := bbPool.Get()
                bb.B = st.MarshalCanonical(bb.B)
                PutStreamTags(st)
                // Calculate the id for the StreamTags
                var sid streamID
                sid.tenantID = tenantID
                sid.id = hash128(bb.B)
                // Store the row
                lr.mustAddInternal(sid, timestamp, fields, bb.B)
                bbPool.Put(bb)
        }

I'm curious if generating stream IDs like this will cause stream id conflicts?

### Troubleshooting docs

- [X] General - https://docs.victoriametrics.com/Troubleshooting.html
- [X] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [X] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by zekker6** [comment]
Hello @acyouzi, stream IDs are based on calculating 128-bit hash of values stored in [stream fields](https://docs.victoriametrics.com/VictoriaLogs/keyConcepts.html#stream-fields).
Those IDs are used in order to uniquely identify a set of similar logs, so it is expected for IDs to be the same for the same set of input fields. The only way which can be treated as a conflict would be hash collision when two different input values produce the same value. While this is very unlikely to happen, it is still possible. In this case log lines which are supposed to be linked to different streams will be merged into one, but no data corruption/crashes will happen.

**Comment by acyouzi** [comment]
> Hello @acyouzi, stream IDs are based on calculating 128-bit hash of values stored in [stream fields](https://docs.victoriametrics.com/VictoriaLogs/keyConcepts.html#stream-fields). Those IDs are used in order to uniquely identify a set of similar logs, so it is expected for IDs to be the same for the same set of input fields. The only way which can be treated as a conflict would be hash collision when two different input values produce the same value. While this is very unlikely to happen, it is still possible. In this case log lines which are supposed to be linked to different streams will be merged into one, but no data corruption/crashes will happen.

@zekker6 Thank you for your reply. One thing I’m confused about is that the index from streamTag to streamId seems to be established only once. [code](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/logstorage/partition.go#L131)
If a hash collision occurs, some indexes from streamTag to streamId will be lost. In this case, wouldn’t there be a problem when using the stream filter to filter logs?

**Comment by zekker6** [comment]
@acyouzi Yes, you are right. In case log entry with different `streamTag` will get the same streamID it won't be registered.

**Comment by acyouzi** [comment]
@zekker6 Okay, so will you make any fixes for this situation in the future?

**Comment by zekker6** [comment]
@acyouzi It is unlikely that there will be a special handling for this case added in the future. There are a few reasons for this:
- Probability of collision occurrence - using 128-bit hash creates a key space with the size of  `2^128`. It is expected that hash will be calculated for each `_stream_fields` combination which [must correspond to one unique logs source](https://docs.victoriametrics.com/VictoriaLogs/keyConcepts.html#stream-fields). Assuming that ["perfect"](https://en.wikipedia.org/wiki/Perfect_hash_function) hash function is used, it will require to have `2^64 = 18_446_744_073_709 551_616` of unique log sources to have a 50% chance of collision (see [birthday attack](https://en.wikipedia.org/wiki/Birthday_attack) for reference).
- In case collision will actually occur the impact would be that when performing a search the response will contain items matching to expected items and ones which appeared as the result of collision.

Based on very low probability of this event and low impact on the result it is unlikely that there will be a logic to handle this event added as of now.

**Comment by jiekun** [comment]
Close as answered.
